url9a01f43c15_poragchowdhury_securitygame_tgz-pJ8-securitygame_GameMasterJ8/src/securitygame/Network.java

```diff
@@ -0,0 +1,533 @@
+				neighbourCount++;
+			}
+		}
+		if(neighbourCount >= Parameters.MAX_ROUTER_EDGES)
+			return;
+		while(neighbourCount < Parameters.MAX_ROUTER_EDGES)
+		{
+			int neighborindex= r.nextInt(nodes.length);
+			if(neighborindex != routerIndex){
+				if(!routerNeighbors.contains(neighborindex))
+				{
+					adjacencyMatrix[routerIndex][neighborindex]=1;
+					adjacencyMatrix[neighborindex][routerIndex]=1;
+					routerNeighbors.add(neighborindex);
+					neighbourCount++;
+				}
+			}
+		}
+	}
+
+	/**
+     * Returns boolean validating a node to be eligible for Neighbor or not
+     * @param currentIndex An integer indicates current node id
+     * @param neighborIndex An integer indicates neighbor node id
+     * @param adjacencyMatrix A two dimensional array for adjacency
+     * @return boolean True/False validating a node to be eligible for Neighbor or not
+     */
+	public boolean isAllowedToBeNeighbor(int currentIndex, int neighborIndex, int [][] adjacencyMatrix)
+	{
+		if (currentIndex == neighborIndex)
+			return false;
+		int neighborCount = 0;
+		for(int i=0; i < adjacencyMatrix[neighborIndex].length; ++i)
+		{
+			if (adjacencyMatrix[neighborIndex][i] == 1)
+				neighborCount++;
+		}
+        return neighborCount < Parameters.MAX_NEIGHBORS;
+
+	}
+
+	/**
+     * Returns size of the network
+     * @return size of the network i.e. number of total nodes
+     */
+	public int getSize()
+	{
+		return nodes.length;
+	}
+
+	/**
+     * Adds Honeypot in the network
+     * @param sv An integer indicates security value
+     * @param pv An integer indicates point value
+     * @param neighbors An integer array indicates all the neighbors
+     */	
+	public void addHoneypot(int sv, int pv, int[]neighbors)
+	{
+		Node[] n = new Node[nodes.length+1];
+		for(int i = 0; i < nodes.length; i++)
+			n[i] = nodes[i];
+		n[nodes.length] = new Node(nodes.length,sv,pv,1);
+
+		for(int i = 0; i < neighbors.length; i++)
+		{
+			n[nodes.length].neighbor.add(nodes[neighbors[i]]);
+			nodes[neighbors[i]].neighbor.add(n[nodes.length]);
+		}
+		nodes = n;
+	}
+
+	/**
+     * Print hidden network in a file.  private or unexplored nodes' rows will have -1.
+     * Edited by Marcus Gutierrez (11/12/2014 - 7:24 AM)
+     */
+	public void printHiddenNetwork()
+	{
+		ArrayList<Node> captured = getCapturedNodes();
+		//System.out.println("Captured: " + captured.size());
+		PrintWriter writer;
+		try {
+			writer = new PrintWriter(fullGraphName + "-hidden.graph", "UTF-8");
+			for (int i = 0; i < nodes.length; i++)
+			{
+				Node node = getNode(i);
+				if (node.isCaptured() == true){
+					int neighborSize = node.neighbor.size();
+
+					int neighborCounter = 0;
+					for(Node neighbor: node.neighbor)
+					{
+						if(neighbor.getNodeID()!=node.getNodeID())
+						{
+							if(neighborCounter==neighborSize-1)
+								writer.print(neighbor.getNodeID());
+							else
+								writer.print(neighbor.getNodeID()+",");
+						}
+						neighborCounter++;
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 199:209, 253:265.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.